corehq/motech/models.py

```diff
@@ -1,20 +1,56 @@
 import re
 
+from django.contrib.postgres.fields import JSONField
 from django.core.serializers.json import DjangoJSONEncoder
 from django.db import models
 
 import jsonfield
 
 from corehq.motech.const import (
     ALGO_AES,
-    BASIC_AUTH,
-    DIGEST_AUTH,
+    AUTH_TYPES,
     OAUTH1,
+    OAUTH2_BEARER,
     PASSWORD_PLACEHOLDER,
 )
 from corehq.motech.utils import b64_aes_decrypt, b64_aes_encrypt
 
 
+class ApiAuthSettings(models.Model):
+    """
+    Stores OAuth1 and OAuth 2.0 endpoints and settings for known APIs.
+    Once an API is added, its settings are available to all domains.
+    """
+    name = models.CharField(max_length=255)  # e.g. "DHIS2"
+    auth_type = models.CharField(
+        max_length=7,
+        choices=(
+            (OAUTH1, "OAuth1"),
+            (OAUTH2_BEARER, "OAuth 2.0 Bearer Tokens"),
+        )
+    )
+    # OAuth1
+    # URL for token to identify HQ. e.g. '/oauth/request_token' (Twitter)
+    request_token_url = models.CharField(max_length=255, null=True, blank=True)
+    # URL for user to authorize HQ. e.g. '/oauth/authorize'
+    authorization_url = models.CharField(max_length=255, null=True, blank=True)
+    # URL to fetch access token. e.g. '/oauth/access_token'
+    access_token_url = models.CharField(max_length=255, null=True, blank=True)
+
+    # OAuth 2.0
+    # URL to fetch bearer token. e.g. '/uaa/oauth/token' (DHIS2)
+    token_url = models.CharField(max_length=255, null=True, blank=True)
+    # URL to refresh bearer token. e.g. '/uaa/oauth/token'
+    refresh_url = models.CharField(max_length=255, null=True, blank=True)
+    # Pass credentials in Basic Auth header when requesting a token?
+    # Otherwise they are passed in the request body.
+    pass_credentials_in_header = models.BooleanField(default=False)
```

Comment:
what about putting all these in a JSON blob? You could define a JsonObject for each different auth type with only the necessary fields.

Reply:
ad750e4